docs/manual/optional-checker.tex

```diff
@@ -98,17 +99,26 @@
 
 \sectionAndLabel{What the Optional Checker guarantees}{optional-guarantees}
 
-The Optional Checker guarantees that your code will not throw an exception
+The Optional Checker guarantees that your code will not throw a \<NoSuchElementException> exception
 due to use of an absent \<Optional> where a present \<Optional> is needed.
 More specifically, the Optional Checker will issue an error if you call
-\sunjavadoc{java.base/java/util/Optional.html\#get()
-}{get}
+\sunjavadoc{java.base/java/util/Optional.html\#get()}{get}
 or
-\sunjavadoc{java.base/java/util/Optional.html\#orElseThrow(java.util.function.Supplier)}{orElseThrow}
+\sunjavadoc{java.base/java/util/Optional.html\#orElseThrow()}{orElseThrow()}
 on a \<@MaybePresent Optional> receiver, because each of these methods
-throws an exception if the receiver is an absent
+throws a \<NoSuchElementException> exception if the receiver is a possibly-absent
 \<Optional>.
 
+By contrast, the Optional Checker does not issue an error if you call
+\sunjavadoc{java.base/java/util/Optional.html\#orElseThrow(java.util.function.Supplier)}{orElseThrow(Supplier)}
+with a possibly-absent \<Optional>.  That method call does not
+% necessarily
+throw \<NoSuchElementException>.  The Optional Checker assumes that the
```

Comment:
I think I understand what you were getting at with the "necessarily" here; what if the user-provided exception supplier throws a `NoSuchElementException`?

I haven't seen this in practice at all; in my opinion, it would be a dual to a programmer explicitly throwing a `NullPointerException`, which is technically legal Java but feels very wrong.

Should the Optional Checker special-case this exception to prevent it from being thrown here? We could also table the discussion until Cameron and I have gone and analyzed more projects.

Reply:
I think that this case will be both rare and bad style, so I don't think we should do anything about it.  That's also why I commented out "necessarily". :-)